### DISCUSSION TITLE
call js from V

### DISCUSSION THREAD

**Discussion by sh4m2hwz** [vlang/v#12682] [Ideas]
Maybe call js code browser or nodejs from v code?
if not, support is planned in this direction?

**Comment by playXE** [comment]
You already can call JS function from V in `.js.v` files:
```v
fn JS.Math.tan(f64) f64

fn tan(x f64) f64 {
    return f64(JS.Math.tan(x)) // do not forget to wrap return value in `f64` again, atm js gen does not do this automatically
}

```

**Comment by sh4m2hwz** [comment]
```vlang
fn JS.Math.tan(f64) f64                 
fn JS.console.log(string)                                                       
fn tan(x f64) f64 {
    return f64(JS.Math.tan(x)) // do not forget to wrap return value in `f64` again, atm js gen does not do this automatically    
}                                                               
JS.console.log("hello from js!")
```
don't work,I'm viewing generated js file and in this file have only this prototype 
```javascript
VV_LOCAL_SYMBOL void main__main(void) {
        JS__console__log(_SLIT("hello drom js!"));
}
VV_LOCAL_SYMBOL void JS__console__log(string );
```
there is some source for calling js from vlang, or what pieces of code you can read in the vlang compiler?

**Comment by playXE** [comment]
@sh4m2hwz you have to compile with `-b js` flag:
```
➜  ~ cat file.js.v
fn JS.Math.tan(f64) f64
fn JS.console.log(string)
fn tan(x f64) f64 {
    return f64(JS.Math.tan(x)) // do not forget to wrap return value in `f64` again, atm js gen does not do this automatically
}
JS.console.log("hello from js!")
➜  ~ v file.js.v -b js
➜  ~ node file.js
{ str: 'hello from js!' }
```

**Comment by sh4m2hwz** [comment]
using the -b option nodejs will not generate console.log ("hello from js!")

**Comment by playXE** [comment]
what do you mean? `-b js` generates JS code compatible with NodeJS. There are 4 options:
1) js_node - NodeJS compatible code
2) js_browser - Browser compatible code
3) js_freestanding - Code compatible with vanilla JS engines (with no runtime)
4) js - alias to js_node

**Comment by KeitoTobi1** [comment]
@playXE 
Would you be willing to support Deno for the future?

**Comment by playXE** [comment]
@KeitoTobi1 yes! Atm I work with NodeJS mainly because it has all its modules enabled by default. AFAIK Deno requires passing CLI arguments to enable IO for example

**Comment by Jengro777** [comment]
```
import requests
import json
import execjs

def get_tk(q):
    with open('google_tk.js', 'r', encoding='utf-8') as f:
        js = f.read()
    j = execjs.compile(js)
    tk = j.call('pu', q)
    return tk

def translate_google():
    q = input("请输入你要翻译的内容：")
    headers = {
        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36",
        "referer": "https://translate.google.cn/"
    }
    url = "https://translate.google.com/translate_a/single?client=webapp&sl=zh-CN&tl=en&hl=zh-CN&dt=at&dt=bd&dt=ex&dt=ld&dt=md&dt=qca&dt=rw&dt=rm&dt=sos&dt=ss&dt=t&otf=2&ssel=0&tsel=0&kc=3&tk=" + get_tk(
        q) + "&q=" + q
    res = requests.get(url, headers=headers)
    data = res.content.decode()
    json_data = json.loads(data)
    print(json_data[0][0][0])

if __name__ == "__main__":
    translate_google()

```

How to use V to call js functions like python

**Comment by Jengro777** [comment]
> Maybe call js code browser or nodejs from v code? if not, support is planned in this direction?

execjs

Is this waiting for a third-party library to implement the function of calling js?

**Comment by Jengro777** [comment]
We hope that Vlang can directly call JavaScript functions, if supported. I hope to provide a more mature case here
